crates/control/src/motion/walking_engine/balancing.rs

```diff
@@ -70,38 +72,65 @@ pub fn step_adjustment(
     backward_foot_support: f32,
     max_adjustment: f32,
     step_adjustment_output: &mut AdditionalOutput<StepAdjustment>,
-) -> (FootOffsets, FootOffsets) {
+    mut left_foot_lift: f32,
+    mut right_foot_lift: f32,
+    stabilization_foot_lift_multiplier: f32,
+    stabilization_foot_lift_offset: f32,
+) -> (FootOffsets, FootOffsets, f32, f32) {
+    let linear_time = (t.as_secs_f32() / planned_step_duration.as_secs_f32()).clamp(0.0, 1.0);
+
     let next_left_forward =
         current_left_foot.forward + next_left_walk_request.forward - last_left_walk_request.forward;
     let next_right_forward = current_right_foot.forward + next_right_walk_request.forward
         - last_right_walk_request.forward;
+
     let backward_balance_limit =
         (next_left_forward).min(next_right_forward) + backward_foot_support;
     let forward_balance_limit = (next_left_forward).max(next_right_forward) + forward_foot_support;
-    let (next_swing_forward, next_support_forward) = match swing_side {
+
+    let (next_swing_foot, next_support_foot) = match swing_side {
         Side::Left => (next_left_forward, next_right_forward),
         Side::Right => (next_right_forward, next_left_forward),
     };
     let adjustment = if torso_tilt_shift < backward_balance_limit {
-        next_swing_forward - torso_tilt_shift - backward_balance_limit
+        -next_swing_foot - torso_tilt_shift.abs() - backward_foot_support.abs()
     } else if torso_tilt_shift > forward_balance_limit {
-        next_swing_forward - torso_tilt_shift - forward_balance_limit
+        -next_swing_foot + torso_tilt_shift.abs() + forward_foot_support.abs()
     } else {
         0.0
     };
+
     let limited_adjustment = adjustment.clamp(-max_adjustment, max_adjustment);
-    let adjusted_swing_forward = next_swing_forward - limited_adjustment;
-    let adjusted_support_forward = next_support_forward + 0.5 * limited_adjustment;
+
+    let adjusted_swing_foot = next_swing_foot + limited_adjustment;
+    let adjusted_support_foot = next_support_foot;
+
+    if adjustment != 0.0 {
+        match swing_side {
+            Side::Left => {
+                left_foot_lift = left_foot_lift * stabilization_foot_lift_multiplier
```

Comment:
You mean like "if the adjustment is != 0, then the swing foot should be lifted higher above the ground"? For example to prevent the swing foot colliding with the ground (because the tip and heel of the swing foot like to get stuck on the ground) too early?

Reply:
That is an idea I tested multiple times already. It is like a double edge sword:

It helps that the swing foot get less stuck when moving forward/backward. On the otherside: if it just holds its position but still has an earlier support foot switch than intended, the robot will "fall" the additional mm difference in height more. Without the height adjustment, the swing foot would have collided with the ground earlier (or with less height change) at the same position. 

I made that observation (I think) when I compared the stability when walking in place (so like 1% forward speed) and pushed the robot backwards. Without the height adjustment the walk was more stable, as the robot would often get the support foot switch when the swing foot has the highest height (so after about 50% of the walk step).

The better solution of course would be to orientate the swing foot perfectly to the ground, as this would prevent the tip and heel from colliding with the ground. But without a good control over the joints and a good enough estimation of the torso orientation this approach would also worsen the stability. (This is one reason why our swing sole orientation is so stupidly complex. In our newest version positive rotation adjustments, e.g. bringing the tip of the sole closer to the ground and the heel further away, uses 50% of the torso orientation. Higher values made it worse again).


**Last but not least**: one idea I tested last year was to increase the interpolation time of the swing height when backwalking. We still use the parabolic function from rUNSWift, but ensure that the highest point is reached at 50% of the stepduration OR after 125 ms (whatever comes first). That way we can increase the duration for the second half, which improves the side walking by a lot already (that way we can side walk with our planned base duration of 250 ms). All without the need to increase the step height itself.

What I added last year was that we do the same when walking backwards too. So when walking backwards the interpolation time from interpolating back to the 0 height of the swing foot takes 12 - 36 ms longer than normal. But keep an eye on the knee temperatures. If I remember correctly, our first version heated up the robots very fast, but this could also have been just one of our worn-out robots as an outlier (it had on both knees 5 degree of play).

Reply:
That's interesting. When I tested the stabilization, increasing the foot height improved the situation as you described, but I think I never noticed the case that the foot fell. But I will investigate this more.

Swing sole orientation prediction sounds highly interesting 🤔

Regarding your last comment: This seems reasonable. We want to probably redo our walking engine next season, we will keep this in mind during design and implementation. In Issue #411 we're currently collecting ideas, feel free to add your comment there. 

Thanks for your input! :)